### PR TITLE
Don't overwrite Base.vect for numbers with a different implementation

### PR DESCRIPTION
**Pull request by ararslan**
Fixes https://github.com/JuliaOpt/Convex.jl/issues/241

### CODE REVIEW

**Comment by iamed2** [comment]
Double-checked that this compiles out the `any` call?

**Comment by iamed2** [comment]
Should we do this with `vcat` as well?

File: src/atoms/affine/stack.jl [hunk 1]
```diff
@@ -132,8 +132,15 @@ function hvcat(rows::Tuple{Vararg{Int}}, args::AbstractExprOrValue...)
...
-Base.vect(args::T...) where {T<:AbstractExpr} = transpose(HcatAtom(map(transpose, args)...))
 Base.vect(args::AbstractExpr...) = transpose(HcatAtom(map(transpose, args)...))
-Base.vect(args::AbstractExprOrValue...) = transpose(HcatAtom(map(arg -> transpose(convert(AbstractExpr, arg)), args)...))
-Base.vect(args::Value...) = collect(args)
+
+function Base.vect(args::AbstractExprOrValue...)
+    if any(_isexpr, args)
+        transpose(HcatAtom(map(arg -> transpose(convert(AbstractExpr, arg)), args)...))
+    else
+        invoke(Base.vect, Tuple{Vararg{Any}}, args...)
```
Comment (mlubin) [inline src/atoms/affine/stack.jl]: Does this break type inference for `Base.vect`?

Comment (ararslan) [reply src/atoms/affine/stack.jl]: Luckily it does not seem to. Using the example in the linked issue:
```julia
julia> @code_warntype [1, 1im]
Body::Array{Complex{Int64},1}
141 1 ── %1  = (Base.getfield)(args, 1, true)::Int64                                                                                  │╻╷╷╷╷ any
    └───       goto #12 if not true                                                                                                   ││┃     _any
    2 ┄─ %3  = φ (#1 => %1, #11 => %19)::Union{Complex{Int64}, Int64}                                                                 │││   
    │    %4  = φ (#1 => 2, #11 => %20)::Int64                                                                                         │││   
    │    %5  = (isa)(%3, Complex{Int64})::Bool                                                                                        │││   
    └───       goto #4 if not %5                                                                                                      │││   
    3 ──       goto #7                                                                                                                │││   
    4 ── %8  = (isa)(%3, Int64)::Bool                                                                                                 │││   
    └───       goto #6 if not %8                                                                                                      │││   
    5 ──       goto #7                                                                                                                │││   
    6 ──       (Core.throw)(ErrorException("fatal error in type inference (type bound)"))                                             │││   
    └───       $(Expr(:unreachable))                                                                                                  │││   
    7 ┄─ %13 = (Base.slt_int)(2, %4)::Bool                                                                                            │││╻╷    iterate
    └───       goto #9 if not %13                                                                                                     ││││  
    8 ──       goto #10                                                                                                               ││││  
    9 ── %16 = (Base.getfield)(args, %4, true)::Union{Complex{Int64}, Int64}                                                          ││││╻     getindex
    │    %17 = (Base.add_int)(%4, 1)::Int64                                                                                           ││││╻     +
    └───       goto #10                                                                                                               ││││  
    10 ┄ %19 = φ (#9 => %16)::Union{Complex{Int64}, Int64}                                                                            │││   
    │    %20 = φ (#9 => %17)::Int64                                                                                                   │││   
    │    %21 = φ (#8 => true, #9 => false)::Bool                                                                                      │││   
    │    %22 = (Base.not_int)(%21)::Bool                                                                                              │││   
    └───       goto #12 if not %22                                                                                                    │││   
    11 ─       goto #2                                                                                                                │││   
    12 ─       goto #13                                                                                                               │││   
    13 ─       goto #14                                                                                                               ││    
    14 ─       goto #16 if not false                                                                                                  │     
    15 ─       nothing                                                                                                                │     
144 16 ─ %29 = Base.vect::Core.Compiler.Const(Base.vect, false)                                                                       │     
    │    %30 = (getfield)(args, 1)::Int64                                                                                             │     
    │    %31 = (getfield)(args, 2)::Complex{Int64}                                                                                    │     
    │    %32 = invoke %29(%30::Int64, %31::Vararg{Any,N} where N)::Array{Complex{Int64},1}                                            │     
    └───       return %32                                                                                                             │     
```

**Comment by ararslan** [comment]
> Double-checked that this compiles out the any call?

It does not, unfortunately, and there is a performance impact on the Base method with this change.

After loading Convex:
```
julia> @benchmark Base.vect($x...)
BenchmarkTools.Trial: 
  memory estimate:  102.72 KiB
  allocs estimate:  375
  --------------
  minimum time:     36.807 μs (0.00% GC)
  median time:      38.912 μs (0.00% GC)
  mean time:        45.703 μs (12.67% GC)
  maximum time:     33.698 ms (99.77% GC)
  --------------
  samples:          10000
  evals/sample:     1
```
(here `x` is a 74-element `Vector{Any}` consisting of `Int`s, `Float32`s, and `Complex{Int}`s.)

Without loading Convex, same `x`:
```
julia> @benchmark Base.vect($x...)
BenchmarkTools.Trial: 
  memory estimate:  55.84 KiB
  allocs estimate:  148
  --------------
  minimum time:     22.036 μs (0.00% GC)
  median time:      23.726 μs (0.00% GC)
  mean time:        28.634 μs (13.89% GC)
  maximum time:     25.832 ms (99.89% GC)
  --------------
  samples:          10000
  evals/sample:     1
```

**Comment by iamed2** [comment]
Can we do it the old way then? I know it's technically type piracy...

**Comment by ararslan** [comment]
Okay, I've reinstated the old code but corrected the implementation.

**Comment by iamed2** [comment]
Hmm what if we changed it to use `invoke` unconditionally?

**Comment by ararslan** [comment]
I checked that earlier and there is still a performance impact, unfortunately. Smaller than with the `any` branch at least, but still slower than without loading Convex. With the current state of the PR, the performance is unchanged after loading Convex (since the code does exactly the same thing).

**Pull request merged**